### PR TITLE
Initial pass at ShouldMatchApproved

### PR DESCRIPTION
**Pull request by JakeGinnivan**
// @dennisroche 

Very basic, but works. Looking for early feedback on extension points/API

### CODE REVIEW

**Comment by JakeGinnivan** [comment]
Am thinking I might pull it out into Shouldly.Approvals

Thoughts? May mean opening up some internals, trying to remove internals visible to

File: src/Shouldly.Shared/DiffToolConfiguration.cs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        readonly List<IShouldNotLaunchDiffTool> _knownShouldNotLaunchDiffToolReasons;
+        readonly List<DiffTool> _diffToolPriority = new List<DiffTool>();
+        readonly List<DiffTool> _diffTools;
+
+        public static class KnownDiffTools
+        {
+            public static readonly DiffTool KDiff3 = new DiffTool("KDiff3", @"C:\Program Files\KDiff3\kdiff3.exe", (received, approved) => $"\"{received}\" \"{approved}\" -o \"{approved}\"");
+        }
+
+        public static class KnownDoNoLaunchStrategies
```
Comment (JosephWoodward) [inline src/Shouldly.Shared/DiffToolConfiguration.cs]: Is this supposed to be DoNotLaunch?

Comment (JakeGinnivan) [inline src/Shouldly.Shared/DiffToolConfiguration.cs]: Yeah =/ :crying_cat_face:

Comment (JosephWoodward) [inline src/Shouldly.Shared/DiffToolConfiguration.cs]: lol sorry, I wasn't trying to be picky or anything!

Comment (cmatskas) [inline src/Shouldly.Shared/DiffToolConfiguration.cs]: Yes, you are! Lol

Sent from Outlook Mobile




On Wed, Dec 30, 2015 at 3:07 AM -0800, "Joseph Woodward" <notifications@github.com> wrote:












In src/Shouldly.Shared/DiffToolConfiguration.cs:
> +using System.Linq;
> +
> +namespace Shouldly
> +{
> +    public class DiffToolConfiguration
> +    {
> +        readonly List<IShouldNotLaunchDiffTool> _knownShouldNotLaunchDiffToolReasons;
> +        readonly List<DiffTool> _diffToolPriority = new List<DiffTool>();
> +        readonly List<DiffTool> _diffTools;
> +
> +        public static class KnownDiffTools
> +        {
> +            public static readonly DiffTool KDiff3 = new DiffTool("KDiff3", @"C:\Program Files\KDiff3\kdiff3.exe", (received, approved) => $"\"{received}\" \"{approved}\" -o \"{approved}\"");
> +        }
> +
> +        public static class KnownDoNoLaunchStrategies



lol sorry, I wasn't trying to be picky or anything!

Comment (JakeGinnivan) [inline src/Shouldly.Shared/DiffToolConfiguration.cs]: Fixed :)

**Comment by JakeGinnivan** [comment]
It's a bit more complete now. Need to add a bunch of different difftools and detection for different CI servers to prevent difftools from popping up. Critical view of the API and naming/structure of things welcome. I plan on basing some of the docs changes on this :)

**Comment by JakeGinnivan** [comment]
// @robdmoore

File: src/Shouldly.Shared/DiffToolConfiguration.cs (hunk 1 above)
Comment (JosephWoodward) [inline src/Shouldly.Shared/DiffToolConfiguration.cs]: :thumbsup: 

lol at @cmatskas stirring the pot ;)

File: src/Shouldly.Shared/DiffToolConfiguration.cs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        readonly List<DiffTool> _diffTools;
+
+        public static class KnownDiffTools
+        {
+            public static readonly DiffTool KDiff3 = new DiffTool("KDiff3", @"C:\Program Files\KDiff3\kdiff3.exe", (received, approved) => $"\"{received}\" \"{approved}\" -o \"{approved}\"");
+        }
+
+        public static class KnownDoNotLaunchStrategies
+        {
+            public static readonly IShouldNotLaunchDiffTool NCrunch = new DoNotLaunchEnvVariable("NCRUNCH");
```
Comment (robdmoore) [inline src/Shouldly.Shared/DiffToolConfiguration.cs]: When you run an individual test in ncrunch does this environment variable still exist?

**Comment by robdmoore** [comment]
I assume that the reason you didn't pull in Approvals is so you can get the string comparison error message semantics from Shouldly?

Obviously there is a risk here in implementing the approvals stuff ourselves when there is a battle-hardened version sitting on the shelf (NuGet server).

File: src/Shouldly.Shared/DiffToolConfiguration.cs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+{
+    public class DiffToolConfiguration
+    {
+        readonly List<IShouldNotLaunchDiffTool> _knownShouldNotLaunchDiffToolReasons;
+        readonly List<DiffTool> _diffToolPriority = new List<DiffTool>();
+        readonly List<DiffTool> _diffTools;
+
+        public static class KnownDiffTools
+        {
+            public static readonly DiffTool KDiff3 = new DiffTool("KDiff3", @"C:\Program Files\KDiff3\kdiff3.exe", (received, approved) => $"\"{received}\" \"{approved}\" -o \"{approved}\"");
```
Comment (robdmoore) [inline src/Shouldly.Shared/DiffToolConfiguration.cs]: https://github.com/approvals/ApprovalTests.Net/tree/535e0f31bc71609100393af1bdeec4e829c6be53/ApprovalTests/Reporters

File: src/Shouldly.Shared/DiffToolConfiguration.cs (hunk 2 above)
Comment (JakeGinnivan) [inline src/Shouldly.Shared/DiffToolConfiguration.cs]: Yep, you can't get the diff tool through nCrunch (same in Approvals)

File: src/Shouldly.Shared/DiffToolConfiguration.cs (hunk 3 above)
Comment (JakeGinnivan) [inline src/Shouldly.Shared/DiffToolConfiguration.cs]: Yep, will be using a few of those if we are happy with overall impl

**Comment by JakeGinnivan** [comment]
Also @simoncropp you may be interested in this. Do you mind me using your BetterAsyncSupport code you wrote for ApprovalTests?

**Comment by SimonCropp** [comment]
I don't mind u using it. But I didn't push for the PR since I realised there are better ways

**Comment by JakeGinnivan** [comment]
Tell me more :P

**Comment by JakeGinnivan** [comment]
@robdmoore There is a number of reasons I wanted to put this into Shouldly. I also discussed this a bit with @dennisroche, I really dislike just reimplementing stuff without a good reason.

1) Very rarely do I need the power/ecosystem that ApprovalTests has, Shouldly will stick to string (maybe object or image) approvals and do it better.
2) There is next to no docs for approvaltests, when I want to add a descriminator like above I have *no* idea how to do it. I know there is a magic attribute I can use but try discovering it. So I implement custom namers. It is really annoying to extend and change. 
3) I don't care about reporters 99% of the time, having to explain to teams about front loaded reporters which run before the reporter you specify because it's running on a build server is confusing. I may introduce reporters similar to ConventionTests or BDDfy which I think is slightly simpler.
4) To fix line ending normalisation properly in ApprovalTests will require me changing quite a few reporters
5) I think I can make this work for uwp and dnx, I don't see approvaltests switching any time soon.
6) Everything is static, so tests cannot run in parallel (which is the default in xunit 2.0)

Pretty much I love ApprovalTests.Net, but I want a simpler and more modern option which fixes my pain points. I have used ApprovalTests enough (and contributed to it) that I think I can take the learnings and ideas.

**Comment by robdmoore** [comment]
Awesome mate! Sounds good to me :)

#excited

**Comment by JakeGinnivan** [comment]
@robdmoore if you get a chance would love your thoughts on stuff I should think about/naming. To give a quick overview:

* Showing Diff is first class, it is currently blacklist for when to not show
* Configuration allows:
  * `.IgnoreLineEndings();`
  * `.WithFileExtension(".cs")`
  * `.NoDiff();` - Forces never showing diff
  * `.WithDescriminator("net40")` - adds the descriminator in the filename, making it easy to do TestName.descriminator.approved.txt
  * `WithStringCompareOptions()` - allows case insensitive
  * `.UseCallerLocation()` - Tells shouldly to use this methods caller for naming. Useful when you have created a test helper
* `ITestMethodFinder` is used for finding the test method for naming. By default it is the first non-shouldly/non-anonymous method in the stacktrace
* Supports finding difftools on the path or in program files (for better chocolatey support)
* `ShouldlyConfiguration.DiffTools.RegisterDiffViewer()` makes it easy for the user to extend
* `ShouldlyConfiguration.DiffTools.SetDiffToolPriorities()` can set your priorities to override the default
* `ShouldlyConfiguration.DiffTools.AddDoNotLaunchStrategy()` easily extend reasons to not launch difftool

Thats it I think. But that covers all the configuration I have needed over time

**Comment by robdmoore** [comment]
I'm happy with all of that - it sounds amazing/simple/discoverable compared to the configuration the options in approvals :)

**Comment by JosephWoodward** [comment]
Is there an option to set the location of the difftool if it differs from the default installation path? I can't seem to find one.

**Pull request merged**